src/core/image.js

```diff
@@ -75,6 +75,74 @@ var PDFImage = (function PDFImageClosure() {
     return dest;
   }
 
+  /**
+   * Downsizes an image mask.
+   * @param {TypedArray} src - The source buffer.
+   * @param {Number} w1 - Original width.
+   * @param {Number} h1 - Original height.
+   * @param {Number} w2 - New width.
+   * @param {Number} h2 - New height.
+   * @returns {Object} Downsized image mask buffer.
+   */
+  function downsizeImageMask(src, w1, h1, w2, h2) {
+    var resizeFactorX = Math.round(w1 / w2);
+    var resizeFactorY = Math.round(h1 / h2);
+    var packedWidth = (w1 + 7) >> 3;
+    var xHop = (resizeFactorX >> 3);
+    xHop = xHop ? xHop : 1;
+    var yHop = resizeFactorY;
+    var bitHop;
+    if (resizeFactorX >= 8) {
+      bitHop = 8;
+    } else if (resizeFactorX >= 4) {
+      bitHop = 4;
+    } else if (resizeFactorX >= 2) {
+      bitHop = 2;
+    } else {
+      bitHop = 1;
+    }
+    var destWidth = (Math.ceil(packedWidth / xHop) * (8 / bitHop) + 7) >> 3;
+    var destHeight = Math.ceil(h1 / yHop);
+    var dest = new Uint8Array(destWidth * destHeight);
+    var resizedWidth = destWidth << 3;
+    var resizedHeight = destHeight;
+    if (resizeFactorX === 1 && resizeFactorY === 1) {
+      dest = src;
+      return { dest, resizedWidth, resizedHeight, };
```

Comment:
If you're passing the `src` through unchanged, I'd assume that you need to also return the *original* width/height as well to prevent rendering errors!?